xdsl/irdl/declarative_assembly_format.py

```diff
@@ -822,21 +822,21 @@ def print(self, printer: Printer, state: PrintingState, op: IRDLOperation) -> No
 class WhitespaceDirective(FormatDirective):
     """
     A whitespace directive, with the following format:
-      whitespace-directive ::= `\n` | ` `
+      whitespace-directive ::= `\n` | ` ` | ``
     This directive is only applied during printing, and has no effect during
     parsing.
     The directive will not request any space to be printed after.
     """
 
-    whitespace: Literal[" ", "\n"]
+    whitespace: Literal[" ", "\n", ""]
     """The whitespace that should be printed."""
 
     def parse(self, parser: Parser, state: ParsingState) -> None:
         pass
 
     def print(self, printer: Printer, state: PrintingState, op: IRDLOperation) -> None:
         printer.print(self.whitespace)
-        state.last_was_punctuation = False
+        state.last_was_punctuation = len(self.whitespace) == 0
```

Comment:
Length of strings is O(n) :)
```suggestion
        state.last_was_punctuation = not bool(self.whitespace)
```

Reply:
This string always has length 0 or 1

Reply:
maybe we could compromise on the simpler ```self.whitespace == ""```?

Reply:
Sounds good